ci-operator/step-registry/openstack/provision/mirror/openstack-provision-mirror-commands.sh

```diff
@@ -0,0 +1,141 @@
+#!/usr/bin/env bash
+
+set -o nounset
+set -o errexit
+set -o pipefail
+
+if [[ "$CONFIG_TYPE" != *"singlestackv6"* ]]; then
+    echo "Skipping step due to CONFIG_TYPE not matching singlestackv6."
+    exit 0
+fi
+
+retry() {
+    local retries=$1
+    local time=$2
+    shift 2
+
+    local count=0
+    until "$@"; do
+      exit=$?
+      count=$(($count + 1))
+      if [ $count -lt $retries ]; then
+        sleep $time
+      else
+        return $exit
+      fi
+    done
+    return 0
+}
+
+export OS_CLIENT_CONFIG_FILE="${SHARED_DIR}/clouds.yaml"
+WORK_DIR=${WORK_DIR:-$(mktemp -d -t shiftstack-ci-XXXXXXXXXX)}
+CLUSTER_NAME=$(<"${SHARED_DIR}"/CLUSTER_NAME)
+CONTROL_PLANE_NETWORK="${CONTROL_PLANE_NETWORK:-$(<"${SHARED_DIR}/CONTROL_PLANE_NETWORK")}"
+BASTION_FLAVOR="${BASTION_FLAVOR:-$(<"${SHARED_DIR}/BASTION_FLAVOR")}"
+BASTION_USER=${BASTION_USER:-cloud-user}
+
+if ! openstack image show $BASTION_IMAGE >/dev/null; then
+		echo "ERROR: Bastion image does not exist: $BASTION_IMAGE"
+		exit 1
+fi
+
+if ! openstack flavor show $BASTION_FLAVOR >/dev/null; then
+		echo "ERROR: Bastion flavor does not exist: $BASTION_FLAVOR"
+		exit 1
+fi
+
+openstack keypair create --public-key ${CLUSTER_PROFILE_DIR}/ssh-publickey mirror-${CLUSTER_NAME}-${CONFIG_TYPE} >/dev/null
+>&2 echo "Created keypair: mirror-${CLUSTER_NAME}-${CONFIG_TYPE}"
+
+sg_id="$(openstack security group create -f value -c id mirror-${CLUSTER_NAME}-${CONFIG_TYPE} \
+  --description "Mirror security group for $CLUSTER_NAME")"
+>&2 echo "Created mirror security group for ${CLUSTER_NAME}: ${sg_id}"
+openstack security group rule create --ingress --protocol tcp --dst-port 22 --description "${CLUSTER_NAME} SSH" "$sg_id" >/dev/null
+openstack security group rule create --ingress --protocol tcp --ethertype IPv6 --remote-ip "::/0" --dst-port 5000 --description "${CLUSTER_NAME} mirror registry" "$sg_id" >/dev/null
+>&2 echo "Created necessary security group rules in ${sg_id}"
+
+server_params="--network $CONTROL_PLANE_NETWORK --image $BASTION_IMAGE --flavor $BASTION_FLAVOR \
+  --security-group $sg_id --key-name mirror-${CLUSTER_NAME}-${CONFIG_TYPE}"
+
+server_id="$(openstack server create -f value -c id $server_params \
+		"mirror-$CLUSTER_NAME-${CONFIG_TYPE}")"
+>&2 echo "Created nova server mirror-${CLUSTER_NAME}-${CONFIG_TYPE}: ${server_id}"
+
+sleep 10
```

Comment:
do we need this sleep here? is it better to add a --wait?

Reply:
you're right, let me do this now